va/validation-authority.go

```diff
@@ -133,25 +143,55 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 	}
 
 	httpRequest.Host = hostName
-	tr := &http.Transport{
-		// We are talking to a client that does not yet have a certificate,
-		// so we accept a temporary, invalid one.
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-		// We don't expect to make multiple requests to a client, so close
-		// connection immediately.
-		DisableKeepAlives: true,
-	}
+	tr := *http.DefaultTransport.(*http.Transport)
+	// We are talking to a client that does not yet have a certificate,
+	// so we accept a temporary, invalid one.
+	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	// We don't expect to make multiple requests to a client, so close
+	// connection immediately.
+	tr.DisableKeepAlives = true
+	// Intercept Dial in order to connect to the IP address we
+	// selected above.
+	defaultDial := tr.Dial
+	tr.Dial = func(_, _ string) (net.Conn, error) {
+		port := "80"
+		if va.TestMode {
+			port = "5001"
+		} else if scheme == "https" {
+			port = "443"
+		}
+		// Ignore the addr selected by net/http.
+		return defaultDial("tcp", net.JoinHostPort(addrs[0].String(), port))
+	}
+	var redirects []string
+	addrUsed := addrs[0]
```

Comment:
This algorithm should also be pulled out into a function so that the other challenges can do the same thing.